core/trino-main/src/test/java/io/trino/sql/planner/optimizations/TestDeterminePartitionCount.java

```diff
@@ -468,4 +488,287 @@ SELECT count(column_a) FROM table_with_stats_a group by column_b
+                        .build(),
+                output(
+                        join(INNER, builder -> builder
+                                .equiCriteria("column_b", "column_b_3")
+                                .left(join(INNER, builder2 -> builder2
+                                        .equiCriteria("column_a", "column_a_0")
+                                        .left(exchange(REMOTE, REPARTITION, FIXED_HASH_DISTRIBUTION, Optional.of(15),
+                                                node(FilterNode.class,
+                                                        tableScan("table_with_stats_a", ImmutableMap.of("column_a", "column_a", "column_b", "column_b")))))
+                                        .right(exchange(LOCAL,
+                                                exchange(REMOTE, REPARTITION, FIXED_HASH_DISTRIBUTION, Optional.of(15),
+                                                        node(DynamicFilterSourceNode.class,
+                                                                tableScan("table_with_stats_b", ImmutableMap.of("column_a_0", "column_a"))))))))
+                                .right(exchange(LOCAL,
+                                        exchange(REMOTE, REPLICATE, FIXED_BROADCAST_DISTRIBUTION, Optional.empty(),
+                                                node(DynamicFilterSourceNode.class,
+                                                        tableScan("small_table_with_stats", ImmutableMap.of("column_b_3", "column_b")))))))));
+    }
+
+    @Test
+    public void testFireWithTaskRetriesArbitraryExchangeOverHashDistributedStage()
+    {
+        @Language("SQL") String query = """
+                SELECT j.column_a FROM
+                   ((SELECT a.column_a, a.column_b FROM table_with_stats_a as a JOIN table_with_stats_b as b ON a.column_a = b.column_a) UNION ALL SELECT * FROM small_table_with_stats s) j
+                   JOIN
+                   small_table_with_stats s2 ON j.column_b = s2.column_b
+                """;
+
+        assertDistributedPlan(
+                query,
+                Session.builder(getQueryRunner().getDefaultSession())
+                        .setSystemProperty(RETRY_POLICY, "task")
+                        .setSystemProperty(FAULT_TOLERANT_EXECUTION_MAX_PARTITION_COUNT, "50")
+                        .setSystemProperty(FAULT_TOLERANT_EXECUTION_MIN_PARTITION_COUNT, "4")
+                        .setSystemProperty(MIN_INPUT_SIZE_PER_TASK, "20MB")
+                        .setSystemProperty(MIN_INPUT_ROWS_PER_TASK, "400")
+                        .setSystemProperty(JOIN_REORDERING_STRATEGY, "none") // keep syntactic order
+                        .build(),
+                output(
+                        join(INNER, builder -> builder
+                                .ignoreEquiCriteria() // criteria uses new symbols output by exchange left exchange which cannot be expressed in plan matcher
```

Comment:
`output by exchange left exchange` -> `output by left exchange`

> cannot be expressed in plan matcher

cannot or exchange matcher just doesn't support that?

Reply:
exchange matcher does not support that and I could not figure out how to add support for that.